### PR TITLE
Simplify organization.public_id

### PR DESCRIPTION
**Pull request by marcospri**
- Store the full ID in the DB, this avoid the need for any extra SQLA fanciness

- Remove Region/RegionService. This seems like an abstraction over the two character region_code setting.

### TODO

- [x] Migration to add `us.org.lms` to existing Organization rows.
- [x] [Change to the report sql code that dealt with public_id vs _public_id ](https://github.com/hypothesis/report/pull/378)
- [x] Manual test admin pages changes.

#### Migration

```docker compose exec postgres psql -U postgres -c "select public_id from organization;"```

```
    public_id        
------------------------
 FFIHD9f1RwyLPfx14zM_7Q
 NISknJi5TgmYk7Ptjq7hfQ
 llxvXeSfSCqL0HY2wSgCWQ
 LDtcl7EUTeW2UERPJLAVtA
 mPgpi06gRSi-hbJ0ZKMcXg
 v_B9qkKqS0qRATuhErygDg
 asbpfjAXTACa9pP6XVLAIA
(7 rows)
```

- ```tox -e dev --run-command 'alembic upgrade head'```

```
docker compose exec postgres psql -U postgres -c "select public_id from organization;"
             public_id             
-----------------------------------
 us.lms.org.FFIHD9f1RwyLPfx14zM_7Q
 us.lms.org.NISknJi5TgmYk7Ptjq7hfQ
 us.lms.org.llxvXeSfSCqL0HY2wSgCWQ
 us.lms.org.LDtcl7EUTeW2UERPJLAVtA
 us.lms.org.mPgpi06gRSi-hbJ0ZKMcXg
 us.lms.org.v_B9qkKqS0qRATuhErygDg
 us.lms.org.asbpfjAXTACa9pP6XVLAIA
(7 rows)
```

- `tox -e dev --run-command 'alembic downgrade -1'        `

```
docker compose exec postgres psql -U postgres -c "select public_id from organization;"
       public_id        
------------------------
 FFIHD9f1RwyLPfx14zM_7Q
 NISknJi5TgmYk7Ptjq7hfQ
 llxvXeSfSCqL0HY2wSgCWQ
 LDtcl7EUTeW2UERPJLAVtA
 mPgpi06gRSi-hbJ0ZKMcXg
 v_B9qkKqS0qRATuhErygDg
 asbpfjAXTACa9pP6XVLAIA
(7 rows)
```

### CODE REVIEW

File: lms/services/organization.py [hunk 1]
```diff
@@ -409,4 +432,5 @@ def service_factory(_context, request) -> OrganizationService:
     return OrganizationService(
         db_session=request.db,
         h_api=request.find_service(HAPI),
+        region_code=request.registry.settings["region_code"],
```
Comment (marcospri) [inline lms/services/organization.py:435]: Inject the necessary value.

File: lms/views/admin/assignment.py [hunk 2]
```diff
@@ -53,9 +53,9 @@ def assignment_dashboard(self):
         response = HTTPFound(
             location=self.request.route_url(
                 "dashboard.assignment",
-                public_id=assignment.groupings[  # noqa: SLF001
+                public_id=assignment.groupings[
```
Comment (marcospri) [inline lms/views/admin/assignment.py:56]: Use the full value, I reckon this was my mistake recently.

**Comment by marcospri** [comment]
Deploying this will break "organization creation" until the migration is run.

We only do that in the admin pages, we can deploy this first thing in the morning.


Report needs to be executed afterwards, after mering: https://github.com/hypothesis/report/pull/378

File: lms/migrations/versions/b512a5cf64ed_public_id_migration.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+revision = "b512a5cf64ed"
+down_revision = "329313b38de1"
+
+
+def upgrade() -> None:
+    conn = op.get_bind()
+    conn.execute(
+        sa.text(
+            """UPDATE "organization" set public_id = 'us.lms.org.' || "public_id";"""
```
Comment (marcospri) [inline lms/migrations/versions/b512a5cf64ed_public_id_migration.py:19]: TODO: this needs  to read region_code or it will change all envs to the same value.

File: lms/models/_mixins/public_id.py [hunk 4]
```diff
@@ -1,66 +0,0 @@
...
-            PublicId(
-                region=_get_current_region(),
-                model_code=self.public_id_model_code,
-                instance_id=self._public_id,
-            )
-        )
-
-    @public_id.inplace.comparator
-    def public_id_comparator(self):
-        return _PublicIdComparator(self._public_id)
```
Comment (seanh) [inline lms/models/_mixins/public_id.py:66]: Just wow. This stuff is _very_ clever and complicated when all that's needed is a normal string column. Definitely good to see this gone. I'm assuming this wasn't doing anything useful other than being an over-complicated way to compare strings

File: lms/models/organization.py [hunk 5]
```diff
@@ -3,20 +3,19 @@
...
     """Model for Organizations comprised of application instances."""
 
     __tablename__ = "organization"
 
-    public_id_model_code = "org"
-    """The short code which identifies this type of model in PublicIdMixin."""
-
     id = sa.Column(sa.Integer(), autoincrement=True, primary_key=True)
 
+    public_id: Mapped[str] = mapped_column(sa.TEXT(), unique=True)
```
Comment (seanh) [inline lms/models/organization.py:17]: So `Organization.public_id` is just a string now :+1:

Is the `sa.TEXT()` here necessary? Doesn't SQLAlchemy infer that from the `Mapped[str]`? (Also it looks like it was `UnicodeText` in the old `PublicIdMixin`.)

File: lms/models/public_id.py [hunk 6]
```diff
@@ -1,100 +0,0 @@
...
-
-from lms.models.region import Region
-
-
-class InvalidPublicId(Exception):
-    """Indicate an error with the specified public id."""
-
-
-@dataclass
-class PublicId:
```
Comment (seanh) [inline lms/models/public_id.py:13]: So not only was there a `PublicIdMixin` but there's also a separate dataclass called `PublicId`? It's so hard to follow! Glad to see this gone as well

File: lms/services/organization.py [hunk 7]
```diff
@@ -58,6 +65,12 @@ def get_by_public_id(self, public_id: str) -> list | None:
 
         :param public_id: Fully qualified public id
         """
+        parts = public_id.split(".")
+
+        if not len(parts) == 4:
```
[the remark this answers is not in the document]
Comment (seanh) [reply lms/services/organization.py:70]: Ok, so `OrganizationService.get_by_public_id` doesn't _have_ to split the `public_id` string (it just needs to compare the string directly to the prefixed strings in the DB), but it's just splitting it so that it can raise this validation error

[the remark this answers is not in the document]
Comment (seanh) [reply lms/services/organization.py:70]: You could easily add this validation to the model as well: https://docs.sqlalchemy.org/en/20/orm/mapped_attributes.html#simple-validators

**Review by seanh — APPROVED** [review]
_Wow_, the way this worked before is one of the worst examples I've seen of over-complicating something that should have been straightforward.

So the way it works now:

* `models.Organization.public_id` is just a simple string column that contains the full ID in the DB, including the `"us.lms.org."` prefix
* `OrganizationService._region_code` holds the value of the `REGION_CODE` envvar, e.g. `"staging"`, `"us"`, or `"ca"`.
* `OrganizationService.get_by_public_id(public_id: str)` just takes a full `public_id` string (including the prefix) and searches for it in the DB. It will also raise an exception if the `public_id` string doesn't look like it has the right prefixed format (but to be honest this probably isn't even necessary?)
* `OrganizationService.create_organization(name)` calls a `_generate_public_id()` helper method to generate the value for `models.Organization.public_id`, something that was previously done with a lot of complicated SQLAlchemy stuff in the model. The model itself can't really generate this value because it doesn't have access to the `REGION_CODE` envvar.

Seems very simple. Did I miss anything?

I wonder how we're gonna deploy this?

The current code expects to find the non-prefixed organization IDs in the DB, and won't behave correctly after the migration has been run and the DB now contains prefixed organization IDs.

On the other the hand, the _new_ code expects to find the prefixed organization IDs in the DB and won't behave correctly _before_ the migration has been run.

**Comment by marcospri** [comment]
> The current code expects to find the non-prefixed organization IDs in the DB, and won't behave correctly after the migration has been run and the DB now contains prefixed organization IDs.

Not necessary, all the code that used organization.public_id did get the full prefix, it was added by python instead of being on the DB.

I might be wrong but the only code that expected it to be un-prefixed was my recent changes to the dashboards, [see this thread ](https://hypothes-is.slack.com/archives/C4K6M7P5E/p1717584404550689) and the code over report that add the prefix there.

File: lms/models/organization.py (hunk 5 above)
Comment (marcospri) [reply lms/models/organization.py:17]: Just the str would generate a migration, added the type to avoid that.
